src/Snap/Internal/Http/Server/Session.hs

```diff
@@ -300,12 +304,36 @@ httpSession !buffer !serverHandler !config !sessionData = loop
     receiveRequest :: IO Request
     receiveRequest = {-# SCC "httpSession/receiveRequest" #-} do
         readEnd' <- Streams.throwIfProducesMoreThan mAX_HEADERS_SIZE readEnd
-        parseRequest readEnd' >>= toRequest
+        (parseRequest readEnd' `E.catch` parseErrHandler) >>= toRequest
+      where
+        parseErrHandler (HttpParseException emsg) = do
+            let msg = mconcat
+                      [ byteString "HTTP/1.1 400 Bad Request\r\n\r\n"
+                      , byteString (S.pack emsg)
+                      , byteString "\r\n"
+                      , flush
+                      ]
+            writeEndB <- mkBuffer
+            Streams.write (Just msg) writeEndB
+            Streams.write Nothing writeEndB
+            terminateSession BadRequestException
     {-# INLINE receiveRequest #-}
 
     --------------------------------------------------------------------------
     toRequest :: IRequest -> IO Request
     toRequest !ireq = {-# SCC "httpSession/toRequest" #-} do
+        -- RFC 7230 section 2.6: "A server can send a 505 (HTTP
+        -- Version Not Supported) response if it wishes, for any
+        -- reason, to refuse service of the client's major protocol
+        -- version."
+        --
+        -- Since HTTP/2 has been released, we *know* that a major
+        -- version larger than 1 is definitely not supported by
+        -- snap-server currently and so it's reasonable to reject such
+        -- doomed to fail requests with the appropriate 505 response
+        -- code early on.
+        when (fst version >= 2) return505
```

Comment:
This part is fine and we can do it without modifying the parsing code. If you want you can reject `fst version /= 1` because we don't support HTTP 0.x either :)